query-algebrizer/src/clauses/mod.rs

```diff
@@ -191,6 +191,48 @@ impl Default for ConjoiningClauses {
     }
 }
 
+/// Cloning.
+impl ConjoiningClauses {
+    fn make_receptacle(&self) -> ConjoiningClauses {
+        let mut concrete = ConjoiningClauses::default();
+        concrete.is_known_empty = self.is_known_empty;
+        concrete.empty_because = self.empty_because.clone();
+
+        concrete.input_variables = self.input_variables.clone();
+        concrete.value_bindings = self.value_bindings.clone();
+        concrete.known_types = self.known_types.clone();
+        concrete.extracted_types = self.extracted_types.clone();
+
+        concrete
+    }
+
+    /// Make a new CC populated with the relevant variable associations in this CC.
+    /// TODO: Note that the CC's table aliaser is not yet usable.
+    fn use_as_template(&self, vars: &BTreeSet<Variable>) -> ConjoiningClauses {
+        let mut template = ConjoiningClauses::default();
+        template.is_known_empty = self.is_known_empty;
+        template.empty_because = self.empty_because.clone();
+
+        template.input_variables = self.input_variables.intersection(vars).cloned().collect();
+        template.value_bindings = intersect_keys(&self.value_bindings, &vars);
+        template.known_types = intersect_keys(&self.known_types, &vars);
+        template.extracted_types = intersect_keys(&self.extracted_types, &vars);
+
+        template
+    }
+}
+
+/// Return a clone of the provided map with only keys that are present in `ks`.
+fn intersect_keys<K: Clone + Ord, V: Clone>(m: &BTreeMap<K, V>, ks: &BTreeSet<K>) -> BTreeMap<K, V> {
+    m.iter()
+     .filter_map(|(k, v)| if ks.contains(k) {
```

Comment:
There would need to be some kind of `filter` in there — it's an intersection, not an extraction. But yeah, there are a bunch of ways to golf this.